embassy-stm32/src/spi/slave.rs

```diff
@@ -0,0 +1,462 @@
+
+        set_txdmaen(T::REGS, true);
+
+        T::REGS.cr1().modify(|w| w.set_spe(true));
+
+        rx_ring_buffer.start();
+        tx_ring_buffer.start();
+
+        SpiSlaveRingBuffered {
+            _inner: self,
+            tx_ring_buffer,
+            rx_ring_buffer,
+        }
+    }
+
+    /// Write a word to the SPI.
+    pub fn write<W: Word>(&mut self, word: W) -> nb::Result<(), Error> {
+        T::REGS.cr1().modify(|w| w.set_spe(true));
+        self.set_word_size(W::CONFIG);
+
+        let _ = transfer_word(T::REGS, word)?;
+
+        Ok(())
+    }
+
+    /// Read a word from the SPI.
+    pub fn read<W: Word>(&mut self) -> nb::Result<W, Error> {
+        T::REGS.cr1().modify(|w| w.set_spe(true));
+        self.set_word_size(W::CONFIG);
+
+        transfer_word(T::REGS, W::default())
+    }
+
+    /// Bidirectionally transfer by writing a word to SPI while simultaneously reading a word from
+    /// the SPI during the same clock cycle.
+    pub fn transfer<W: Word>(&mut self, word: W) -> nb::Result<W, Error> {
+        T::REGS.cr1().modify(|w| w.set_spe(true));
+        self.set_word_size(W::CONFIG);
+
+        transfer_word(T::REGS, word)
+    }
+}
+
+#[cfg(not(gpdma))]
+impl<'d, T: Instance, W> SpiSlaveRingBuffered<'d, T, W>
+where
+    W: Word,
+{
+    /// Write elements from `buf` into the transmit ringbuffer. These elements will be transmitted
+    /// over SPI in the background using DMA.
+    ///
+    /// The number of elements that were read from `buf` is returned. An overrun error occurs when
+    /// the portion to be written to was read by DMA.
+    pub fn write(&mut self, buf: &[W]) -> Result<usize, Error> {
+        // `WritableRingBuffer` errors with Overrun if we try to send an empty buffer
+        if buf.is_empty() {
+            return Ok(0);
+        }
+
+        match self.tx_ring_buffer.write(buf) {
+            Ok((written, _)) => Ok(written),
+            Err(_) => {
+                self.tx_ring_buffer.clear();
+                Err(Error::Overrun)
+            }
+        }
+    }
+
+    /// Read elements from the receive ringbuffer into `buf`. Elements received over SPI are
+    /// written into the receive ringbuffer in the background using DMA.
+    ///
+    /// The number of elements that were written into `buf` is returned. An overrun error occurs
+    /// when the portion to be read was overwritten by DMA.
+    pub fn read(&mut self, buf: &mut [W]) -> Result<usize, Error> {
+        match self.rx_ring_buffer.read(buf) {
+            Ok((read, _)) => Ok(read),
+            Err(_) => {
+                self.rx_ring_buffer.clear();
+                Err(Error::Overrun)
+            }
+        }
+    }
+
+    /// Read an exact number of elements from the receive ringbuffer into `buf`.
+    ///
+    /// An overrun error occurs when the portion to be read was overwritten by DMA.
+    pub async fn read_exact(&mut self, buf: &mut [W]) -> Result<(), Error> {
```

Comment:
It would probably be good to note in the docstring that this triggers only when the DMA buffer is half full (it's documented in the DMA documentation, but probably it'd be nice to mention it here and in the other async methods as well - I ran into this and missed this when first testing).